### DISCUSSION TITLE
Single OPFS file virtual file system (maybe using whio_epfs by @sgbeal)

### DISCUSSION THREAD

**Discussion by samwillis** [rhashimoto/wa-sqlite#165] [Ideas]
An idea occurred to me, would it be possible to implement either a SQLite VFS or Emscripten WASMFS that mounted a single OPFS file that internally folds multiple files. That way it would be possible to asynchronously "mount" the file system on start then have full sync access to open and create additional files. Essentially it's a re-implementation of a low level filesystem to work around the restrictions of OPFS.

While looking into this I came across [whio_epfs](https://fossil.wanderinghorse.net/r/whio/wiki?name=whio_epfs) by @sgbeal, I'm wandering if there is a route to an epfs based Emscripten WASMFS...

For SQLite it's less of a problem as the access handle pool VFS doesn't need to hold onto more than a handful of pre-opened access handles. But for other software compiled to WASM (such as Postgres) there can be many more files needed, and the AHP trick could get out of hand.

Has anyone considered this approach already?

Obviously the downside is that it could opened by a single worker exclusively, but that's the same trade of as AHPVFS

**Comment by sgbeal** [comment]
> While looking into this I came across whio_epfs...

Oh, wow, a blast from the far past. That's one of those projects i'd always like to return to if i could just find some use for it. (It and its parent i/o library were written for fun, not because of an acute need for them.)

> maybe using whio_epfs...

epfs has always been missing one important component which would be very sorely needed for any serious usage: some sort of journaling. Back when epfs and its predecessor were written i was not confident that i could pull that off properly, so never bit that proverbial bullet.

> ... wandering if there is a route to an epfs based Emscripten WASMFS...

FWIW, i see the following problems with using WASMFS as a basis, based on experience with use of it as a stepping stone early on in the sqlite project's OPFS adventure:

- It's perpetually flagged as "in development" and has changed in incompatible ways out from under us before (tickling the collective aversion to 3rd-party dependencies which is coincidentally shared by all members of the sqlite project).
- It didn't work on ARM platforms last i checked (but it's been a good 6 months or more since i have).
- It's specific to a single vendor. If they abandon it or go another direction with it, all clients are left hanging. Google backs emscripten and Google's history with regards to pulling the rug out from under developers of its libraries [is still a sore point for me](https://groups.google.com/g/v8-users/c/MUq5WrC2kcE).

That said:

- The concept is of course sound and, at least in native space, proven to work (sans something akin to sqlite's journaling support, which would be a necessity in the face of browser tab closures, power-cycling mobile devices, etc.). However, i'm instinctively unconvinced (as distinct from "convinced by demonstrated facts") that things like rollback journals can be implemented as reliably on OPFS as they can in native code, for very the simple reason that i don't trust browsers across all wasm-targeted platforms to fully fsync() when JS tells them to.

- Hypothetically, WASMFS wouldn't strictly be needed, though something like Emscripten's "FS" API probably(?) would: the FS API would enable one to hide a virtual FS inside a single epfs-like container file in OPFS, which could be mounted just like WASMFS can.

**Comment by rhashimoto** [comment]
> Has anyone considered this approach already?

I did [consider it](https://github.com/rhashimoto/wa-sqlite/discussions/67#discussioncomment-4249212) as a conceptual proof...

> I believe it is possible to build a fully general and synchronous SQLite VFS with the new OPFS spec, one that would work without Asyncify and without SharedArrayBuffer. The basic concept is instead of exposing the OPFS file system to SQLite, you implement a file system using OPFS as a device. The simplest example is to use a single OPFS file as a block device and port a real file system on top of it, inodes and all that.

...but I didn't pursue it because I don't know much about real file systems.

> Obviously the downside is that it could opened by a single worker exclusively, but that's the same trade of as AHPVFS

Note that this restriction is [no longer in Chrome](https://chromestatus.com/feature/5172892632875008), and Firefox has taken a [positive position](https://github.com/mozilla/standards-positions/issues/861) on the change.
